libretro-common/file/file_path.c

```diff
@@ -650,6 +661,105 @@ bool path_is_absolute(const char *path)
    return false;
 }
 
+bool is_posix_path(const char* path)
+{
+   if (strchr(path, posix_path_delimiter))
+      return true;
+
+   return false;
+}
+
+bool is_windows_path(const char* path)
+{
+   if (strchr(path, posix_path_delimiter))
+      return false;
+
+   return true;
+}
+
+/**
+ * path_resolve_to_local_file_system:
+ * @buf                    : output buffer
+ * @path                   : original relative path
+ * @base_content_directory : base content directory
+ *
+ * Resolves @path to local file system, switching '\' with '/' if needed (and viceversa)
+ * Appends @path to @base_content_directory
+ *
+ * Example[1]
+ *  @path = 'roms\zelda3.zip'
+ *  @base_content_directory = '/retroarch/'
+ *
+ *  Under linux/android returns '/retroarch/roms/zelda3.zip'
+ *
+ * Example[2]
+ *  @path = 'roms\zelda3.zip'
+ *  @base_content_directory = 'c:\games\retroarch\'
+ *
+ *  Under windows returns 'c:\games\retroarch\roms\zelda3.zip'
+ *
+ * Example[3]
+ *  @path = 'roms/zelda3.zip'
+ *  @base_content_directory = 'c:\games\retroarch\'
+ *
+ *  Under windows returns 'c:\games\retroarch\roms\zelda3.zip'
+ **/
+void path_resolve_to_local_file_system(char* buf, const char* path, const char* base_content_directory, size_t count)
```

Comment:
Again, this is not quite generalised enough for `file_path.c`

- It needs to be made very clear that this will only work if `path` is relative

- `base_content_directory` is playlist terminology - this should be `in_refpath` (in fact, I'd make the arguments: `char *out_path, const char *in_path, const char *in_refpath, size_t size`, so it matches `fill_pathname_resolve_relative()`)

- `base_content_directory` is never converted... It assumes the reference path is always in local format (which we cannot do in general)

- This function can error out - but there's no return value to indicate when this happens...

Reply:
The function tries to convert a input relative path to the local file system, appending it to another input refpath (if defined). 
It EXPECTS the input path to be absolute, if not it leaves it unchanged and exits: it does not mean that it failed, but simply that the input path is already absolute, and there is no need for conversion, so we just keep it as it is.

Conversion of `\` to `/` and viceversa currently applies only to the relative path stored in playlists, which we agreed to always be stored in unix format (for example '/roms/zelda3.zip'). 
The base absolute path (the rgui_browser_directory config parameter) is expected to already be in the local file system format: changing 'C:\roms' to 'C:/roms' under unix is useless, the same changing '/roms/' to '\roms\' under Windows.

For the same reason, I do not translate `\` to `/` if the input path is already absolute:
- if input path is a windows path (for example "C:\roms\zelda3.zip"), under windows we don't need to change it, under unix it wouldn't even work (C:/roms/zelda3.zip ??). 
- The same if the input path is a posix path: '/roms/zelda3.zip' can only be used under unix, on windows is unusable ('\roms\zelda3.zip' ??).

So what we do in this function is to merge a relative input path (in posix format) to a base absolute path (already in the local file system format). The relative path is first changed to local format (only if needed), and then merged with the base absolute path. 

The function itself does not fail, I guess the returned path may be invalid for the local fs, but for that I relied on the file load functions to fail and return an error.

Also, the function is currently called only when loading playlists, and only if relative paths are enabled, which means the   rgui_browser_directory parameter was selected in retroarch (so already in the local file system format, the only way to break it would be to manually edit the config file..).

Reply:
Maybe it's better if I move this function to another file?
For example, I could add a new `playlist_relative_path.c` file with this function, and all related defines, so it's purpose is clearer?

Reply:
Ah ha - you initially missed my point, but then you got it:

> Maybe it's better if I move this function to another file?

Yes, exactly :)

The function does what it needs to do, but it's just not *generic* enough for `libretro-common`. So yes, you could either move it to a separate 'playlist path' library, or (more simply, perhaps) just place it in `playlist.c`

Reply:
Ok I moved the functions to `playlist.c`.